deps/rabbit/src/rabbit_fifo.erl

```diff
@@ -604,293 +601,189 @@ apply(Meta, {nodeup, Node}, #?MODULE{consumers = Cons0,
-               consumer_strategy = rabbit_fifo_v1:get_cfg_field(consumer_strategy, V1State),
-               delivery_limit = rabbit_fifo_v1:get_cfg_field(delivery_limit, V1State),
-               expires = rabbit_fifo_v1:get_cfg_field(expires, V1State)
-              },
-
-    MessagesConsumersV2 = maps:fold(fun(_ConsumerId, #consumer{checked_out = Checked}, Acc) ->
-                                            Acc + maps:size(Checked)
-                                    end, 0, ConsumersV2),
-    MessagesWaitingConsumersV2 = lists:foldl(fun({_ConsumerId, #consumer{checked_out = Checked}}, Acc) ->
-                                                     Acc + maps:size(Checked)
-                                             end, 0, WaitingConsumersV2),
-    MessagesTotal = lqueue:len(MessagesV2) +
-                    lqueue:len(ReturnsV2) +
-                    MessagesConsumersV2 +
-                    MessagesWaitingConsumersV2,
-
-    #?MODULE{cfg = Cfg,
-             messages = MessagesV2,
-             messages_total = MessagesTotal,
-             returns = ReturnsV2,
-             enqueue_count = rabbit_fifo_v1:get_field(enqueue_count, V1State),
-             enqueuers = EnqueuersV2,
-             ra_indexes = IndexesV1,
-             release_cursors = rabbit_fifo_v1:get_field(release_cursors, V1State),
-             consumers = ConsumersV2,
-             service_queue = rabbit_fifo_v1:get_field(service_queue, V1State),
-             msg_bytes_enqueue = rabbit_fifo_v1:get_field(msg_bytes_enqueue, V1State),
-             msg_bytes_checkout = rabbit_fifo_v1:get_field(msg_bytes_checkout, V1State),
-             waiting_consumers = WaitingConsumersV2,
-             last_active = rabbit_fifo_v1:get_field(last_active, V1State)
-            }.
-
-convert_v2_to_v3(#rabbit_fifo{consumers = ConsumersV2} = StateV2) ->
-    ConsumersV3 = maps:map(fun(_, C) ->
-                                   convert_consumer_v2_to_v3(C)
-                           end, ConsumersV2),
-    StateV2#rabbit_fifo{consumers = ConsumersV3}.
-
-convert_consumer_v2_to_v3(C = #consumer{cfg = Cfg = #consumer_cfg{credit_mode = simple_prefetch,
-                                                                  meta = #{prefetch := Prefetch}}}) ->
-    C#consumer{cfg = Cfg#consumer_cfg{credit_mode = {simple_prefetch, Prefetch}}};
-convert_consumer_v2_to_v3(C) ->
-    C.
+convert_v3_to_v4(#{} = _Meta, StateV3) ->
+    %% TODO: consider emitting release cursors as checkpoints
+    Messages0 = rabbit_fifo_v3:get_field(messages, StateV3),
+    Returns0 = lqueue:to_list(rabbit_fifo_v3:get_field(returns, StateV3)),
+    Consumers0 = rabbit_fifo_v3:get_field(consumers, StateV3),
+    Consumers = maps:map(
+                  fun (_, #consumer{checked_out = Ch0} = C) ->
+                          Ch = maps:map(
+                                 fun (_, ?MSG(I, #{delivery_count := DC} = H)) ->
+                                         ?MSG(I, H#{acquired_count => DC});
```

Comment:
Not too important, but I think deleting `delivery_count` makes sense (in addition to adding `acquired_count`) because then the new `delivery_count` is only ever set when the AMQP 1.0 client rejects or modifies the message (which is the new v4 behaviour).